### DISCUSSION TITLE
Need some help testing `useFormContext` and `useWatch`

### DISCUSSION THREAD

**Discussion by ghostling** [react-hook-form/react-hook-form#2802] [Q&A] [answered]
Hi, I have a nested form component that is utilizing react-hook-form in this manner:
```
function ParentForm() {
  const formMethods = useForm();

  return (
    <FormProvider {...formMethods}>
      <ChildForm />
    </FormProvider>
  )
}

function ChildForm() {
  const {control} = useFormContext();
  const watch = useWatch({ name: 'someField' });

  return (...);
}
```

And when I try to test `<ChildForm />` with React Testing Library, I'm unsure how to wrap the `<ChildForm />` in a test form provider. I've tried to use some mocking libraries like:
```
import {mock, when} from 'ts-mockito';
const mockMethods: FormProviderProps = mock<FormProviderProps>();

describe('<ChildForm />', () => {
  it('should render expected contents', () => {
    const { getByTestId, getByText } = render(
      <FormProvider {...mockMethods}>
        <ChildForm />,
      </FormProvider>
    );
  });
});
```

I get an error:
```
Error: Uncaught [TypeError: Cannot destructure property 'watchFieldsHookRef' of '(control || methods.control)' as it is undefined.]
```
which I see probably comes from [here](https://github.com/react-hook-form/react-hook-form/blob/03216edb0b29bae631fb752e6ec612111d82798d/src/useWatch.ts#L57).

Also tried something similar with [testdouble td.object](https://github.com/testdouble/testdouble.js). Constructing the entire fake object manually seems really tedious, so that is not an option.

I looked at the useWatch.test and useFormContext.test in the repo and they seem to test those hooks themselves rather than a component that calls those hooks... Any suggestions on how to accomplish this?  Thank you.

**Comment by bluebill1049** [comment, marked as answer]
don't mock those methods, just do `useForm()`

**Reply by ghostling** [reply]
I don't understand how exactly, can you provide an example? I get this error when trying to `useForm()` in a test body:
```
    Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
    1. You might have mismatching versions of React and the renderer (such as React DOM)
    2. You might be breaking the Rules of Hooks
    3. You might have more than one copy of React in the same app
    See https://fb.me/react-invalid-hook-call for tips about how to debug and fix this problem.
```

**Reply by bluebill1049** [reply]
take a look at our tests, there are plenty of examples in the repo 👍

**Reply by ghostling** [reply]
Actually I manage to get something like this working: 
```
    const WrapperForm = ({ children }: any) => {
      const formMethods = useForm();
      return (
        <FormProvider {...formMethods}>
          {children}
        </FormProvider>);
    };

    const { getByTestId, getByText } = render(
      <WrapperForm>
        <ChildForm />
      </WrapperForm>
    );
```
Hopefully I won't have to duplicate too much of the Parent Form functionality into test itself. 😓

**Reply by adamduncan** [reply]
I'm running into a possibly related issue, trying to test a fieldset that implements `useFormContext` and `watch`, which is a child component of a form set up with `FormProvider`:

```jsx
// FieldsetExample.js
const { control, watch } = useFormContext();
const { append, remove } = useFieldArray({
  control,
  name: 'appliances',
});

const appliances = watch('appliances');

//... render/register fields
```

Where in my tests I'm asserting against a label in the UI that's based on `appliances` field value:

```js
import { render, screen } from '@testing-library/react';
import { FormProvider, useForm } from 'react-hook-form';

const FormProviderWrapper = ({ children }) => {
  const form = useForm({
    appliances: [],
  });
  return <FormProvider {...form}>{children}</FormProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: FormProviderWrapper, ...options });

describe(`Example fieldset component`, () => {
  beforeEach(() => {
    customRender(<FieldsetExample />);
  });

  it('renders empty state', () => {
    expect(screen.queryByText(/0 appliances/i)).toBeInTheDocument();
  });
});
```

I'm seeing `appliances` show up as `undefined` as the test runs, which fails any assertions that depend on it or a property/method call of it (e.g. `appliances.length`).

Is there something I'm missing here?
